atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/DynamicTaskScheduler.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.sweep.asts;
+
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.refreshable.Refreshable;
+import java.time.Duration;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+// This feels like it should already exist.
+public final class DynamicTaskScheduler {
+    private static final SafeLogger log = SafeLoggerFactory.get(DynamicTaskScheduler.class);
+    private final ScheduledExecutorService scheduledExecutorService;
+    private final Refreshable<Duration> automaticSweepRefreshDelay;
+    private final Runnable task;
+    private final AtomicBoolean isStarted = new AtomicBoolean(false);
+
+    private DynamicTaskScheduler(
+            ScheduledExecutorService scheduledExecutorService,
+            Refreshable<Duration> automaticSweepRefreshDelay,
+            Runnable task) {
+        this.scheduledExecutorService = scheduledExecutorService;
+        this.automaticSweepRefreshDelay = automaticSweepRefreshDelay;
+        this.task = task;
+    }
+
+    public static DynamicTaskScheduler create(
+            ScheduledExecutorService scheduledExecutorService,
+            Refreshable<Duration> automaticSweepRefreshDelay,
+            Runnable task) {
+        return new DynamicTaskScheduler(scheduledExecutorService, automaticSweepRefreshDelay, task);
+    }
+
+    public void start() {
+        if (isStarted.compareAndSet(false, true)) {
+            scheduleNextIteration(automaticSweepRefreshDelay.get());
+        } else {
+            log.warn("Attempted to start an already started task");
+        }
+    }
+
+    private void runOneIteration() {
+        Duration delay = automaticSweepRefreshDelay.get();
+        try {
+            log.info("Running task");
```

Comment:
Should we provide some kind of task ID (see `safeLoggablePurpose` in atlasdb proxy)? Otherwise if this is reused I imagine the logs will get confusing